vm/src/builtins/genericalias.rs

```diff
@@ -0,0 +1,166 @@
+    }
+}
+
+#[pyimpl(with(Hashable), flags(BASETYPE))]
+impl PyGenericAlias {
+    pub fn new(origin: PyTypeRef, args: PyObjectRef, vm: &VirtualMachine) -> Self {
+        let args: PyTupleRef = if let Ok(tuple) = PyTupleRef::try_from_object(vm, args.clone()) {
+            tuple
+        } else {
+            vm.ctx
+                .new_tuple(vec![args])
+                .downcast::<PyTuple>()
+                .ok()
+                .unwrap()
+        };
+
+        Self {
+            origin,
+            args: args.clone(),
+            parameters: make_parameters(args, vm),
+        }
+    }
+
+    #[pymethod(magic)]
+    fn repr(&self, vm: &VirtualMachine) -> PyResult<String> {
+        fn repr_item(obj: PyObjectRef, vm: &VirtualMachine) -> PyResult<String> {
+            if obj.is(&vm.ctx.ellipsis) {
+                return Ok("...".to_string());
+            }
+
+            if vm.get_attribute_opt(obj.clone(), "__origin__")?.is_some()
+                && vm.get_attribute_opt(obj.clone(), "__args__")?.is_some()
+            {
+                return Ok(vm.to_repr(&obj)?.as_str().to_string());
+            }
+
+            match (
+                vm.get_attribute_opt(obj.clone(), "__qualname__")?
+                    .and_then(|o| o.downcast_ref::<PyStr>().map(|n| n.as_str().to_string())),
+                vm.get_attribute_opt(obj.clone(), "__module__")?
+                    .and_then(|o| o.downcast_ref::<PyStr>().map(|m| m.as_str().to_string())),
+            ) {
+                (None, _) | (_, None) => Ok(vm.to_repr(&obj)?.as_str().to_string()),
+                (Some(qualname), Some(module)) => Ok(if module == "builtins" {
+                    qualname
+                } else {
+                    format!("{}.{}", module, qualname)
+                }),
+            }
+        }
+
+        Ok(format!(
+            "{}[{}]",
+            repr_item(self.origin.as_object().clone(), vm)?,
+            self.args
+                .as_slice()
+                .iter()
+                .map(|o| repr_item(o.clone(), vm))
+                .collect::<PyResult<Vec<_>>>()?
+                .join(", ")
+        ))
+    }
+
+    #[pyproperty(magic)]
+    fn parameters(&self) -> PyObjectRef {
+        self.parameters.as_object().clone()
+    }
+
+    #[pyproperty(magic)]
+    fn args(&self) -> PyObjectRef {
+        self.args.as_object().clone()
+    }
+
+    #[pyproperty(magic)]
+    fn origin(&self) -> PyObjectRef {
+        self.origin.as_object().clone()
+    }
+}
+
+fn is_typevar(obj: PyObjectRef) -> bool {
+    let class = obj.class();
+    class.tp_name() == "TypeVar"
+        && match class.get_attr("__module__") {
+            Some(o) => o
+                .downcast_ref::<PyStr>()
+                .map(|s| s.as_str() == "typing")
+                .unwrap_or(false),
+            None => false,
+        }
+}
+
+fn make_parameters(args: PyTupleRef, vm: &VirtualMachine) -> PyTupleRef {
```

Comment:
```suggestion
fn make_parameters(args: &PyTupleRef, vm: &VirtualMachine) -> PyTupleRef {
```

Because it doesn't use any tuple feature, taking slice is also an option

```suggestion
fn make_parameters(args: &[PyObjectRef], vm: &VirtualMachine) -> PyTupleRef {
```